espnet2/enh/layers/iva.py

```diff
@@ -0,0 +1,513 @@
+    W: torch.Tensor, shape (..., n_frequencies, n_channels, n_channels), optional
+        The demixing matrix, it is updated if provided
+    A: torch.Tensor, shape (..., n_frequencies, n_channels, n_channels), optional
+        The mixing matrix, it is updated if provided
+
+    Returns
+    -------
+    X: torch.Tensor, shape (n_frequencies, n_channels, n_frames)
+        The updated source estimates
+    """
+    n_chan, n_freq, n_frames = X.shape[-3:]
+
+    # Update now the demixing matrix
+    for s in range(n_chan):
+        v_num = (
+            torch.einsum(
+                "...cfn,...cfn,...fn->...cf", X, weights, X[..., s, :, :].conj()
+            )
+            / n_frames
+        )
+        v_denom = (
+            torch.einsum("...cfn,...fn->...cf", weights, mag_sq(X[..., s, :, :]))
+            / n_frames
+        )
+
+        v = v_num / torch.clamp(v_denom, min=eps)
+        v_s = 1.0 - (1.0 / torch.sqrt(torch.clamp(v_denom[..., s, :], min=eps)))
+        v = torch.cat((v[..., :s, :], v_s[..., None, :], v[..., s + 1 :, :]), dim=-2)
+
+        # update demixed signals
+        X = X - torch.einsum("...cf,...fn->...cfn", v, X[..., s, :, :])
+
+        if W is not None:
+            W = W - torch.einsum("...cf,...fd->...fcd", v, W[..., s, :])
+
+        if A is not None:
+            u = torch.einsum("...fcd,...df->...fc", A, v) / torch.clamp(
+                1.0 - v_s[..., None], min=eps
+            )
+            A = torch.cat(
+                (A[..., :s], A[..., [s]] + u[..., None], A[..., s + 1 :]), dim=-1
+            )
+
+    return X, W, A
+
+
+def spatial_model_update_ip2(
+    Xo: torch.Tensor,
+    weights: torch.Tensor,
+    W: Optional[torch.Tensor] = None,
+    A: Optional[torch.Tensor] = None,
+    eps: Optional[float] = 1e-7,
+):
+    """
+    Apply the spatial model update via the generalized eigenvalue decomposition.
+    This method is specialized for two channels.
+
+    Parameters
+    ----------
+    Xo: torch.Tensor, shape (..., n_frequencies, n_channels, n_frames)
+        The microphone input signal with n_chan == 2
+    weights: torch.Tensor, shape (..., n_frequencies, n_channels, n_frames)
+        The weights obtained from the source model to compute
+        the weighted statistics
+
+    Returns
+    -------
+    X: torch.Tensor, shape (n_frequencies, n_channels, n_frames)
+        The updated source estimates
+    """
+    assert Xo.shape[-3] == 2, "This method is specialized for two channels processing."
+
+    V = []
+    for k in [0, 1]:
+        # shape: (n_batch, n_freq, n_chan, n_chan)
+        Vloc = torch.einsum(
+            "...fn,...cfn,...dfn->...fcd", weights[..., k, :, :], Xo, Xo.conj()
+        )
+        Vloc = Vloc / Xo.shape[-1]
+        # make sure V is hermitian symmetric
+        Vloc = 0.5 * (Vloc + hermite(Vloc))
+        V.append(Vloc)
+
+    eigval, eigvec = eigh_2x2(V[1], V[0], eps=eps)
+
+    # reverse order of eigenvectors
+    eigvec = torch.flip(eigvec, dims=(-1,))
```

Comment:
For better compatibility with older PyTorch versions, it it better to replace this line with
```python
    eigvec = reverse(eigvec, dim=-1)
```

and add the following lines to the top:
```python
from espnet2.enh.layers.complex_utils import reverse
```